synapseclient/core/upload/multipart_upload_async.py

```diff
@@ -0,0 +1,884 @@
+    )
+    async def _fetch_pre_signed_part_urls_async(
+        self,
+        upload_id: str,
+        part_numbers: List[int],
+    ) -> Mapping[int, str]:
+        trace.get_current_span().set_attributes({"synapse.upload_id": upload_id})
+        response = await post_file_multipart_presigned_urls(
+            upload_id=upload_id,
+            part_numbers=part_numbers,
+            synapse_client=self._syn,
+        )
+
+        part_urls = {}
+        for part in response["partPresignedUrls"]:
+            part_urls[part["partNumber"]] = (
+                part["uploadPresignedUrl"],
+                part.get("signedHeaders", {}),
+            )
+
+        return part_urls
+
+    def _refresh_pre_signed_part_urls(
+        self,
+        part_number: int,
+        expired_url: str,
+    ) -> Tuple[str, Dict[str, str]]:
+        """Refresh all unfetched presigned urls, and return the refreshed
+        url for the given part number. If an existing expired_url is passed
+        and the url for the given part has already changed that new url
+        will be returned without a refresh (i.e. it is assumed that another
+        thread has already refreshed the url since the passed url expired).
+
+        Arguments:
+            part_number: the part number whose refreshed url should
+                         be returned
+            expired_url: the url that was detected as expired triggering
+                         this refresh
+        Returns:
+            refreshed URL
+
+        """
+        with tracer.start_as_current_span(
+            "UploadAttemptAsync::_refresh_pre_signed_part_urls"
+        ):
+            with self._thread_lock:
+                current_url, headers = self._pre_signed_part_urls[part_number]
+                if current_url != expired_url:
+                    # if the url has already changed since the given url
+                    # was detected as expired we can assume that another
+                    # thread already refreshed the url and can avoid the extra
+                    # fetch.
+                    refreshed_url = current_url, headers
+                else:
+                    self._pre_signed_part_urls = wrap_async_to_sync(
+                        self._fetch_pre_signed_part_urls_async(
+                            self._upload_id,
+                            list(self._pre_signed_part_urls.keys()),
+                        ),
+                        syn=self._syn,
+                    )
+
+                    refreshed_url = self._pre_signed_part_urls[part_number]
+
+            return refreshed_url
+
+    async def _handle_part_wrapper(self, part_number: int) -> HandlePartResult:
+        loop = asyncio.get_running_loop()
+        otel_context = context.get_current()
+
+        mem_info = psutil.virtual_memory()
+
+        if mem_info.available <= self._part_size * 2:
+            gc.collect()
```

Comment:
I'm surprised garbage collection wasn't part of the description in the PR.  Can you add a description as to why you did this?

Reply:
I'm not sure if I really like having this logic in the code and I might get rid of it or change it to only look at the lowest generation for GC.

The problem I am solving:
Each part (8MB) of a file is loaded into memory before it's sent to the storage provider. In testing on a 1GB EC2 instance, I kept killing the machine for out-of-memory issues. What I believe the issue was is that there was not enough items (700 items, 25 items, 10 items IIRC for each generation level) to cause a the scheduled garbage collection.

When I had tested this logic it seemed to have worked - though it was all very inconsistent.